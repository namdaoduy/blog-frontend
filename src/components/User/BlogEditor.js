import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import 'draft-js/dist/Draft.css';
import React, { Component } from 'react';
import ReactMde from 'react-mde';
import 'react-mde/lib/styles/css/react-mde-all.css';
import { connect } from 'react-redux';
import * as Showdown from 'showdown';
import { createBlog, getEdittingBlog, updateBlog } from '../../actions/user';
import '../../assets/styles/user.css';
import * as constants from '../../constants/common';
import history from '../../utils/history';
import Header from '../Common/Header';

class BlogEditor extends Component {
  constructor(props) {
    super(props);
    const blog_id = props.match.params.blog_id;
    this.state = {
      id: blog_id || null,
      type: blog_id ? 'edit' : 'new',
      title: '',
      body: 'Your body here',
      isLoading: true,
      error: false,
    };

    this.converter = new Showdown.Converter({
      tables: true,
      simplifiedAutoLink: true,
    });
  }

  componentDidMount() {
    const { id, type } = this.state;
    if (type === 'new') {
      this.setState({
        isLoading: false,
        title: '',
        body: 'Your body here',
      });
    } else {
      this.fetchBlog(id);
    }
  }

  handleBodyChange = (body) => {
    this.setState({ body });
  }

  handleTitleChange = (e) => {
    this.setState({ title: e.target.value });
  }

  // If create new blog
  handleNew = (_data) => {
    const { title, body } = _data;
    const data = { title, body };
    this.props.createBlog(data)
      .then((res) => {
        console.log(res.success);
        history.push('/user');
      })
      .catch(err => console.log(err));
  }

  // If edit an existed blog
  handleEdit = (data) => {
    this.props.updateBlog(data)
      .then((res) => {
        console.log(res.success);
        history.push('/user');
      })
      .catch(err => console.log(err));
  }

  // Check if input is valid
  validateInput = () => {
    const { title, body } = this.state;
    if (title.length < constants.BLOG_TITLE_LENGTH_MIN
      || title.length > constants.BLOG_TITLE_LENGTH_MAX
      || body.length < constants.BLOG_BODY_LENGTH_MIN
      || body.length > constants.BLOG_BODY_LENGTH_MAX) {
      this.setState({ error: true });
      return false;
    }
    return true;
  }

  // Submit form
  onSubmit = () => {
    if (!this.validateInput()) {
      return false;
    }
    const { id, title, body, type } = this.state;
    const data = { id, title, body };
    if (type === 'edit') {
      this.handleEdit(data);
    } else if (type === 'new') {
      this.handleNew(data);
    }
    return true;
  }

  fetchBlog = (id) => {
    this.props.getEdittingBlog(id)
      .then((res) => {
        if (!res.success) return;
        const { edittingBlog } = this.props;
        this.setState({
          isLoading: false,
          title: edittingBlog.title,
          body: edittingBlog.body,
        });
      })
      .catch(err => console.log(err));
  }

  renderEditor = () => {
    const { isLoading, body } = this.state;
    if (isLoading) {
      return null;
    }
    return (
      <ReactMde
        className="md-editor"
        onChange={this.handleBodyChange}
        value={body}
        generateMarkdownPreview={markdown => Promise.resolve(this.converter.makeHtml(markdown))}
      />
    );
  }

  // render error if input is not valid
  renderError = () => {
    const { error } = this.state;
    if (error) {
      return (
        <div style={{ color: 'red' }}>
          {`Title must be between ${constants.BLOG_TITLE_LENGTH_MIN}-${constants.BLOG_TITLE_LENGTH_MAX} characters.
          Body must be between ${constants.BLOG_BODY_LENGTH_MIN}-${constants.BLOG_BODY_LENGTH_MAX} characters`}
        </div>
      );
    }
    return null;
  }

  render() {
    const { title } = this.state;
    return (
      <div className="user-container">
        <Header />
        <div className="user-inner">
          <div className="editor-container">
            <TextField
              onChange={this.handleTitleChange}
              value={title}
              className="editor-title-input"
              placeholder="10 - 100 characters"
              required
              id="outlined-required"
              label="Title"
              margin="normal"
              variant="outlined"
            />

            <Button
              variant="contained"
              color="secondary"
              onClick={this.onSubmit}
            >
              Submit
            </Button>

            {this.renderError()}

            {this.renderEditor()}

          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = ({ user }) => ({
  edittingBlog: user.edittingBlog,
});

const mapDispatchToProps = {
  createBlog,
  updateBlog,
  getEdittingBlog,
};

export default connect(mapStateToProps, mapDispatchToProps)(BlogEditor);