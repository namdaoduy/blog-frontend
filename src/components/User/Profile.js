import React, { Component } from 'react';
import { Typography, Button, Grid, Divider, Avatar } from '@material-ui/core';
import Paper from '@material-ui/core/Paper';
import DeleteForever from '@material-ui/icons/DeleteForever';
import BorderColor from '@material-ui/icons/BorderColor';
import Moment from 'react-moment';
import '../../assets/styles/user.css';
import { MuiThemeProvider } from '@material-ui/core/styles';
import Header from '../Common/Header';
import theme from '../../constants/theme';
import API from '../../services/apis';
import history from '../../utils/history';

export default class Profile extends Component {
  constructor(props) {
    super(props);
    this.state = {
      user_id: localStorage.getItem('justblog_user_id') || null,
      blogs: [],
      user_info: {},
    };
  }

  componentDidMount() {
    const uid = localStorage.getItem('justblog_user_id') || null;
    if (uid) {
      this.setState({ user_id: uid }, () => {
        this.fetchUserInfo();
        setTimeout(() => {
          this.fetchUserBlogs();
        }, 1000);
      });
    }
  }

  fetchUserInfo = () => {
    const { user_id } = this.state;
    API.getUserInfo(user_id)
      .then((res) => {
        this.setState({ user_info: res.data });
      })
      .catch(err => console.log(err));
  }

  fetchUserBlogs = () => {
    const { user_id } = this.state;
    API.getBlogsByUser(user_id)
      .then((res) => {
        this.setState({ blogs: res.data });
      })
      .catch(err => console.log(err));
  }

  handleDelete = (blog_id) => {
    const del = window.confirm('Are you sure to delete?');
    if (!del) return;
    const token = localStorage.getItem('justblog_access_token');
    API.deleteBlogById(blog_id, token)
      .then(() => {
        window.location.reload();
      })
      .catch(err => console.log(err));
  }

  handleClickEdit = (id) => {
    history.push(`/user/blog/${id}`);
  }

  handleClickNew = () => {
    history.push('/user/blog/new');
  }

  renderBlogs = () => {
    const { blogs } = this.state;
    return blogs.slice(0).reverse().map((blog, i) => (
      <Paper className="user-profile-blog">
        <Typography variant="h5" className="serif-2">
          {blog.title || 'Blog title'}
        </Typography>
        <Typography variant="body1" noWrap>
          {blog.body || 'Blog body'}
        </Typography>
        <Typography variant="subtitle2">
          {blog.author || 'Blog author'}
        </Typography>
        <Typography variant="caption">
          <Moment fromNow>{blog.created_at || 1545813100264}</Moment>
          <span className="dot-divider" />
          {blog.timeRead || `${10} min read`}
        </Typography>
        <Button
          color="secondary"
          variant="fab"
          onClick={() => this.handleDelete(blog.id)}
        >
          <DeleteForever />
        </Button>
        <Button
          color="secondary"
          variant="fab"
          onClick={() => this.handleClickEdit(blog.id)}
        >
          <BorderColor />
        </Button>
      </Paper>
    ));
  }

  render() {
    const { user_info } = this.state;
    const { name, email, picture } = user_info;
    return (
      <MuiThemeProvider theme={theme}>
        <div className="user-container">
          <Header />
          <Grid
            className="user-inner"
            container
            spacing={24}
          >
            <Grid item xs={3}>
              <Avatar alt="User Avatar" className="user-avatar" src={picture || null} />
              <Typography variant="h5" className="bold">
                {name || 'User Name'}
              </Typography>
              <Typography variant="subtitle1">
                {email || 'useremail@gmail.com'}
              </Typography>
              <Button
                variant="contained"
                color="secondary"
                onClick={this.handleClickNew}
              >
                New Blog
              </Button>
            </Grid>

            <Grid item xs={9}>
              <Typography variant="h5" className="semi-bold serif-2" gutterBottom>
                Your Blogs
              </Typography>
              <Divider />
              {this.renderBlogs()}
            </Grid>
          </Grid>
        </div>
      </MuiThemeProvider>
    );
  }
}