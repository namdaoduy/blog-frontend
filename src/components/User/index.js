import React, { Component } from 'react';
import '../../assets/styles/user.css';
import { Switch, Route } from 'react-router-dom';
import { Grid } from '@material-ui/core';
import { MuiThemeProvider } from '@material-ui/core/styles';
import Header from '../Common/Header';
import theme from '../../constants/theme';
import Profile from './Profile';
import BlogEditor from './BlogEditor';
import API from '../../services/apis';

export default class User extends Component {
  constructor(props) {
    super(props);
    this.state = {
      user_id: localStorage.getItem('justblog_user_id'),
      blogs: [],
    };
  }

  componentDidMount() {
    this.fetchUserBlogs();
  }

  fetchUserBlogs = () => {
    API.getBlogsByUser(this.state.user_id)
      .then((res) => {
        this.setState({ blogs: res.data });
      });
  }

  handleDelete = (id) => {
    const del = window.confirm('Are you sure to delete?');
    if (!del) return;
    const token = localStorage.getItem('justblog_access_token');
    API.deleteBlogById(id, token)
      .then((res) => {
        window.location.reload();
      })
      .catch(err => console.log(err));
  }

  handleClickEdit = (id) => {
    this.props.history.push(`/user/blog/${id}`);
  }

  handleClickNew = () => {
    this.props.history.push('/user/blog/new');
  }

  handleNew = (data) => {
    const token = localStorage.getItem('justblog_access_token');
    API.postBlog(data.title, data.body, token)
      .then((res) => {
        console.log(res);
        this.props.history.push('/user');
      });
  }

  handleEdit = (data) => {
    const token = localStorage.getItem('justblog_access_token');
    API.putBlog(data.id, data.title, data.body, token)
      .then((res) => {
        console.log(res);
        this.props.history.push('/user');
      });
  }

  render() {
    return (
      <MuiThemeProvider theme={theme}>
        <div className="user-container">
          <Header />
          <Grid
            className="user-inner"
            container
            spacing={24}
          >
            <Switch>
              <Route
                path="/user/blog/new"
                render={props => (
                  <BlogEditor
                    {...props}
                    type="new"
                    key={`blogNew${Date.now()}`}
                    handleNew={this.handleNew.bind(this)}
                  />
                )}
              />
              <Route
                path="/user/blog/:blog_id"
                render={props => (
                  <BlogEditor
                    {...props}
                    type="edit"
                    key={`blogEdit${Date.now()}`}
                    handleEdit={this.handleEdit.bind(this)}
                  />
                )}
              />
              <Route
                path="/user"
                render={props => (
                  <Profile
                    {...props}
                    key={`profile${Date.now()}`}
                    handleClickNew={this.handleClickNew.bind(this)}
                    handleClickEdit={this.handleClickEdit.bind(this)}
                    handleDelete={this.handleDelete}
                    handleEdit={this.handleEdit}
                    blogs={this.state.blogs}
                  />
                )}
              />
            </Switch>
          </Grid>
        </div>
      </MuiThemeProvider>
    );
  }
}