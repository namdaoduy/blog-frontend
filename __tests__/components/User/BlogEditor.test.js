import React from 'react';
import { shallow } from 'enzyme';
import { BlogEditor } from 'components/User/BlogEditor';

describe('components/User/Profile', () => {
  let wrapper;
  let props;

  const setup = () => {
    wrapper = shallow(<BlogEditor {...props} />);
  };

  beforeEach(() => {
    props = {
      match: { params: { blog_id: 1 } },
      edittingBlog: {
        id: 1,
        title: 'Mock title',
        body: 'Mock body',
        author: 'Mock Author',
        created_at: 123456789,
      },
      createBlog: jest.fn(async () => {}),
      updateBlog: jest.fn(async () => {}),
      getEdittingBlog: jest.fn(async () => {}),
    };
  });

  describe('Test render', () => {
    it('Render correctly', () => {
      setup();
      expect(wrapper).toMatchSnapshot();
    });

    it('Render error correctly', () => {
      setup();
      wrapper.setState({ error: true });
      expect(wrapper).toMatchSnapshot();
    });

    it('Render editor correctly', () => {
      setup();
      wrapper.setState({ isLoading: false });
      expect(wrapper).toMatchSnapshot();
    });

    it('Render if create new blog', () => {
      props.match.params = {};
      setup();
      expect(wrapper).toMatchSnapshot();
    });
  });
});