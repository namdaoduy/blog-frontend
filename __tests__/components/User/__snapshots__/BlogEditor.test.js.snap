// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/User/Profile Test render Render correctly 1`] = `
<div
  className="user-container"
>
  <Connect(Header) />
  <div
    className="user-inner"
  >
    <div
      className="editor-container"
    >
      <TextField
        className="editor-title-input"
        id="outlined-required"
        label="Title"
        margin="normal"
        onChange={[Function]}
        placeholder="10 - 100 characters"
        required={true}
        select={false}
        value=""
        variant="outlined"
      />
      <WithStyles(Button)
        color="secondary"
        onClick={[Function]}
        variant="contained"
      >
        Submit
      </WithStyles(Button)>
    </div>
  </div>
</div>
`;

exports[`components/User/Profile Test render Render editor correctly 1`] = `
<div
  className="user-container"
>
  <Connect(Header) />
  <div
    className="user-inner"
  >
    <div
      className="editor-container"
    >
      <TextField
        className="editor-title-input"
        id="outlined-required"
        label="Title"
        margin="normal"
        onChange={[Function]}
        placeholder="10 - 100 characters"
        required={true}
        select={false}
        value=""
        variant="outlined"
      />
      <WithStyles(Button)
        color="secondary"
        onClick={[Function]}
        variant="contained"
      >
        Submit
      </WithStyles(Button)>
      <ReactMde
        buttonContentOptions={
          Object {
            "iconProvider": [Function],
          }
        }
        className="md-editor"
        commands={
          Array [
            Object {
              "commands": Array [
                Object {
                  "buttonContentBuilder": [Function],
                  "buttonProps": Object {
                    "aria-label": "Add header",
                  },
                  "children": Array [
                    Object {
                      "buttonContentBuilder": [Function],
                      "execute": [Function],
                    },
                    Object {
                      "buttonContentBuilder": [Function],
                      "execute": [Function],
                    },
                    Object {
                      "buttonContentBuilder": [Function],
                      "execute": [Function],
                    },
                    Object {
                      "buttonContentBuilder": [Function],
                      "execute": [Function],
                    },
                  ],
                },
                Object {
                  "buttonContentBuilder": [Function],
                  "buttonProps": Object {
                    "aria-label": "Add bold text",
                  },
                  "execute": [Function],
                  "keyCommand": "bold",
                },
                Object {
                  "buttonContentBuilder": [Function],
                  "buttonProps": Object {
                    "aria-label": "Add italic text",
                  },
                  "execute": [Function],
                  "keyCommand": "italic",
                },
                Object {
                  "buttonContentBuilder": [Function],
                  "buttonProps": Object {
                    "aria-label": "Add strikethrough text",
                  },
                  "execute": [Function],
                },
              ],
            },
            Object {
              "commands": Array [
                Object {
                  "buttonContentBuilder": [Function],
                  "buttonProps": Object {
                    "aria-label": "Insert a link",
                  },
                  "execute": [Function],
                },
                Object {
                  "buttonContentBuilder": [Function],
                  "buttonProps": Object {
                    "aria-label": "Insert a quote",
                  },
                  "execute": [Function],
                },
                Object {
                  "buttonContentBuilder": [Function],
                  "buttonProps": Object {
                    "aria-label": "Insert code",
                  },
                  "execute": [Function],
                  "keyCommand": "code",
                },
                Object {
                  "buttonContentBuilder": [Function],
                  "buttonProps": Object {
                    "aria-label": "Insert a picture",
                  },
                  "execute": [Function],
                },
              ],
            },
            Object {
              "commands": Array [
                Object {
                  "buttonContentBuilder": [Function],
                  "buttonProps": Object {
                    "aria-label": "Insert a bulleted list",
                  },
                  "execute": [Function],
                },
                Object {
                  "buttonContentBuilder": [Function],
                  "buttonProps": Object {
                    "aria-label": "Insert numbered list",
                  },
                  "execute": [Function],
                },
                Object {
                  "buttonContentBuilder": [Function],
                  "buttonProps": Object {
                    "aria-label": "Insert checklist",
                  },
                  "execute": [Function],
                },
              ],
            },
          ]
        }
        emptyPreviewHtml="<p>&nbsp;</p>"
        generateMarkdownPreview={[Function]}
        l18n={
          Object {
            "preview": "Preview",
            "write": "Write",
          }
        }
        maxEditorHeight={500}
        minEditorHeight={200}
        minPreviewHeight={200}
        onChange={[Function]}
        readOnly={false}
        value="Your body here"
      />
    </div>
  </div>
</div>
`;

exports[`components/User/Profile Test render Render error correctly 1`] = `
<div
  className="user-container"
>
  <Connect(Header) />
  <div
    className="user-inner"
  >
    <div
      className="editor-container"
    >
      <TextField
        className="editor-title-input"
        id="outlined-required"
        label="Title"
        margin="normal"
        onChange={[Function]}
        placeholder="10 - 100 characters"
        required={true}
        select={false}
        value=""
        variant="outlined"
      />
      <WithStyles(Button)
        color="secondary"
        onClick={[Function]}
        variant="contained"
      >
        Submit
      </WithStyles(Button)>
      <div
        style={
          Object {
            "color": "red",
          }
        }
      >
        Title must be between 10-100 characters.
          Body must be between 1000-20000 characters
      </div>
    </div>
  </div>
</div>
`;

exports[`components/User/Profile Test render Render if create new blog 1`] = `
<div
  className="user-container"
>
  <Connect(Header) />
  <div
    className="user-inner"
  >
    <div
      className="editor-container"
    >
      <TextField
        className="editor-title-input"
        id="outlined-required"
        label="Title"
        margin="normal"
        onChange={[Function]}
        placeholder="10 - 100 characters"
        required={true}
        select={false}
        value=""
        variant="outlined"
      />
      <WithStyles(Button)
        color="secondary"
        onClick={[Function]}
        variant="contained"
      >
        Submit
      </WithStyles(Button)>
      <ReactMde
        buttonContentOptions={
          Object {
            "iconProvider": [Function],
          }
        }
        className="md-editor"
        commands={
          Array [
            Object {
              "commands": Array [
                Object {
                  "buttonContentBuilder": [Function],
                  "buttonProps": Object {
                    "aria-label": "Add header",
                  },
                  "children": Array [
                    Object {
                      "buttonContentBuilder": [Function],
                      "execute": [Function],
                    },
                    Object {
                      "buttonContentBuilder": [Function],
                      "execute": [Function],
                    },
                    Object {
                      "buttonContentBuilder": [Function],
                      "execute": [Function],
                    },
                    Object {
                      "buttonContentBuilder": [Function],
                      "execute": [Function],
                    },
                  ],
                },
                Object {
                  "buttonContentBuilder": [Function],
                  "buttonProps": Object {
                    "aria-label": "Add bold text",
                  },
                  "execute": [Function],
                  "keyCommand": "bold",
                },
                Object {
                  "buttonContentBuilder": [Function],
                  "buttonProps": Object {
                    "aria-label": "Add italic text",
                  },
                  "execute": [Function],
                  "keyCommand": "italic",
                },
                Object {
                  "buttonContentBuilder": [Function],
                  "buttonProps": Object {
                    "aria-label": "Add strikethrough text",
                  },
                  "execute": [Function],
                },
              ],
            },
            Object {
              "commands": Array [
                Object {
                  "buttonContentBuilder": [Function],
                  "buttonProps": Object {
                    "aria-label": "Insert a link",
                  },
                  "execute": [Function],
                },
                Object {
                  "buttonContentBuilder": [Function],
                  "buttonProps": Object {
                    "aria-label": "Insert a quote",
                  },
                  "execute": [Function],
                },
                Object {
                  "buttonContentBuilder": [Function],
                  "buttonProps": Object {
                    "aria-label": "Insert code",
                  },
                  "execute": [Function],
                  "keyCommand": "code",
                },
                Object {
                  "buttonContentBuilder": [Function],
                  "buttonProps": Object {
                    "aria-label": "Insert a picture",
                  },
                  "execute": [Function],
                },
              ],
            },
            Object {
              "commands": Array [
                Object {
                  "buttonContentBuilder": [Function],
                  "buttonProps": Object {
                    "aria-label": "Insert a bulleted list",
                  },
                  "execute": [Function],
                },
                Object {
                  "buttonContentBuilder": [Function],
                  "buttonProps": Object {
                    "aria-label": "Insert numbered list",
                  },
                  "execute": [Function],
                },
                Object {
                  "buttonContentBuilder": [Function],
                  "buttonProps": Object {
                    "aria-label": "Insert checklist",
                  },
                  "execute": [Function],
                },
              ],
            },
          ]
        }
        emptyPreviewHtml="<p>&nbsp;</p>"
        generateMarkdownPreview={[Function]}
        l18n={
          Object {
            "preview": "Preview",
            "write": "Write",
          }
        }
        maxEditorHeight={500}
        minEditorHeight={200}
        minPreviewHeight={200}
        onChange={[Function]}
        readOnly={false}
        value="Your body here"
      />
    </div>
  </div>
</div>
`;
